src/datatype/functions_vecf32.rs

```diff
@@ -0,0 +1,295 @@
+        if is_null {
+            None
+        } else {
+            let ptr = NonNull::new(datum.cast_mut_ptr::<AccumulateStateHeader>()).unwrap();
+            unsafe { Some(AccumulateState::new(ptr)) }
+        }
+    }
+}
+
+impl IntoDatum for AccumulateState<'_> {
+    fn into_datum(self) -> Option<Datum> {
+        Some(Datum::from(self.into_raw() as *mut ()))
+    }
+
+    fn type_oid() -> Oid {
+        let namespace = pgrx::pg_catalog::PgNamespace::search_namespacename(c"vectors").unwrap();
+        let namespace = namespace.get().expect("pgvecto.rs is not installed.");
+        let t = pgrx::pg_catalog::PgType::search_typenamensp(c"accumulate_state", namespace.oid())
+            .unwrap();
+        let t = t.get().expect("pg_catalog is broken.");
+        t.oid()
+    }
+}
+
+unsafe impl SqlTranslatable for AccumulateState<'_> {
+    fn argument_sql() -> Result<SqlMapping, ArgumentError> {
+        Ok(SqlMapping::As(String::from("accumulate_state")))
+    }
+    fn return_sql() -> Result<Returns, ReturnsError> {
+        Ok(Returns::One(SqlMapping::As(String::from(
+            "accumulate_state",
+        ))))
+    }
+}
+
+fn parse_accumulate_state(input: &[u8]) -> Result<(u64, Vec<F32>), String> {
+    use crate::utils::parse::parse_vector;
+    let hint = "Invalid input format for accumulatestate, using \'bigint, array \' like \'1, [1]\'";
+    let (count, slice) = input.split_once(|&c| c == b',').ok_or(hint)?;
+    let count = std::str::from_utf8(count)
+        .map_err(|e| e.to_string() + "\n" + hint)?
+        .parse::<u64>()
+        .map_err(|e| e.to_string() + "\n" + hint)?;
+    let v = parse_vector(slice, 0, |s| s.parse().ok());
+    match v {
+        Err(e) => Err(e.to_string() + "\n" + hint),
+        Ok(vector) => Ok((count, vector)),
+    }
+}
+
+#[pgrx::pg_extern(immutable, strict, parallel_safe)]
+fn _vectors_accumulate_state_in(input: &CStr, _oid: Oid, _typmod: i32) -> AccumulateState<'_> {
+    // parse one bigint and a vector of f32, split with a comma
+    let res = parse_accumulate_state(input.to_bytes());
+    match res {
+        Err(e) => {
+            bad_literal(&e.to_string());
+        }
+        Ok((count, vector)) => AccumulateState::new_with_slice(count, &vector),
+    }
+}
+
+#[pgrx::pg_extern(immutable, strict, parallel_safe)]
+fn _vectors_accumulate_state_out(state: AccumulateState<'_>) -> CString {
+    let mut buffer = String::new();
+    buffer.push_str(format!("{}", state.count()).as_str());
+    buffer.push('[');
+    if let Some(&x) = state.slice().first() {
+        buffer.push_str(format!("{}", x).as_str());
+    }
+    for &x in state.slice().iter().skip(1) {
+        buffer.push_str(format!(", {}", x).as_str());
+    }
+    buffer.push(']');
+    CString::new(buffer).unwrap()
+}
+
+/// accumulate intermediate state for vector average
+#[pgrx::pg_extern(immutable, strict, parallel_safe)]
+fn _vectors_vector_accum<'a>(
```

Comment:
I meet unused-lifetimes warning here. I don't configure out the reason here. As far as I know , this warning only occurs with #[pgrx::pg_extern].
The 235th line is the same.

Reply:
The macro expands to
```rust
pub unsafe extern "C" fn _vectors_bvecf32_send_wrapper<'a>(
    _fcinfo: ::pgrx::pg_sys::FunctionCallInfo,
) -> ::pgrx::pg_sys::Datum
```
The lifetime parameter is redundant here. You could use `allow` to suppress this warning before pgrx fixes it.

Reply:
I just discovered this problem too. After add `#![allow(unused_lifetimes)] #![allow(clippy::extra_unused_lifetimes)]` in file header, it works. Thanks!